File: Sources/HoverPocket/PocketApps/PocketAppHealthStore.swift
import Darwin
import Foundation

enum PocketAppHealthStatus: String, Codable, Sendable {
    case healthy
    case attention
    case unused
    case disabled
}

struct PocketAppHealthSnapshot: Equatable, Sendable {
    let packageID: String
    let status: PocketAppHealthStatus
    let reasonCode: String
    let lastUsedAt: Date?
    let lastSuccessfulActivationAt: Date?
    let consecutiveActivationFailures: Int
    let disableSuggested: Bool
}

enum PocketAppHealthError: Error, Equatable {
    case invalid
    case storage
    case readback
}

final class PocketAppHealthStore {
    private struct Record: Codable, Equatable {
        let recordVersion: Int
        let packageID: String
        var firstActivatedAt: Date?
        var lastSuccessfulActivationAt: Date?
        var lastUsedAt: Date?
        var lastFailureAt: Date?
        var consecutiveActivationFailures: Int
        var updatedAt: Date
    }

    static let unusedInterval: TimeInterval = 30 * 24 * 60 * 60
    private static let usageWriteInterval: TimeInterval = 5 * 60
    private static let maximumRecordBytes = 16 * 1_024

    private let rootDirectory: URL

    init(rootDirectory: URL) throws {
        self.rootDirectory = rootDirectory.standardizedFileURL
        do {
            try FileManager.default.createDirectory(
                at: self.rootDirectory,
                withIntermediateDirectories: true,
                attributes: [.posixPermissions: 0o700]
            )
            try requireSafeRoot()
        } catch let error as PocketAppHealthError {
            throw error
        } catch {
            throw PocketAppHealthError.storage
        }
    }

    func recordActivationSuccess(packageID: String, now: Date = Date()) throws {
        try update(packageID: packageID, now: now) { record in
            if record.firstActivatedAt == nil { record.firstActivatedAt = now }
            record.lastSuccessfulActivationAt = now
            record.consecutiveActivationFailures = 0
        }
    }

    func recordActivationFailure(packageID: String, now: Date = Date()) throws {
        try update(packageID: packageID, now: now) { record in
            record.lastFailureAt = now
            record.consecutiveActivationFailures = min(record.consecutiveActivationFailures + 1, 1_000)
        }
    }

    func recordUse(packageID: String, now: Date = Date()) throws {
        guard Self.validPackageID(packageID) else { throw PocketAppHealthError.invalid }
        if let current = try read(packageID: packageID),
           let lastUsedAt = current.lastUsedAt,
           now.timeIntervalSince(lastUsedAt) >= 0,
           now.timeIntervalSince(lastUsedAt) < Self.usageWriteInterval {
            return
        }
        try update(packageID: packageID, now: now) { record in
            if record.firstActivatedAt == nil { record.firstActivatedAt = now }
            record.lastUsedAt = now
        }
    }

    func snapshots(
        packages: [PocketAppManagedPackage],
        issues: [PocketAppManagementIssue],
        now: Date = Date()
    ) -> [PocketAppHealthSnapshot] {
        var result: [PocketAppHealthSnapshot] = []
        for package in packages where package.state != .removed {
            do {
                let record = try read(packageID: package.packageID)
                result.append(Self.snapshot(package: package, record: record, now: now))
            } catch {
                result.append(PocketAppHealthSnapshot(
                    packageID: package.packageID,
                    status: .attention,
                    reasonCode: "HEALTH_METADATA_CORRUPT",
                    lastUsedAt: nil,
                    lastSuccessfulActivationAt: nil,
                    consecutiveActivationFailures: 0,
                    disableSuggested: false
                ))
            }
        }
        for issue in issues {
            let record = try? read(packageID: issue.packageID)
            result.append(PocketAppHealthSnapshot(
                packageID: issue.packageID,
                status: .attention,
                reasonCode: issue.errorCode,
                lastUsedAt: record?.lastUsedAt,
                lastSuccessfulActivationAt: record?.lastSuccessfulActivationAt,
                consecutiveActivationFailures: record?.consecutiveActivationFailures ?? 0,
                disableSuggested: false
            ))
        }
        return result.sorted { $0.packageID < $1.packageID }
    }

    private func update(
        packageID: String,
        now: Date,
        mutate: (inout Record) -> Void
    ) throws {
        guard Self.validPackageID(packageID), now.timeIntervalSince1970.isFinite else {
            throw PocketAppHealthError.invalid
        }
        var record = try read(packageID: packageID) ?? Record(
            recordVersion: 1,
            packageID: packageID,
            firstActivatedAt: nil,
            lastSuccessfulActivationAt: nil,
            lastUsedAt: nil,
            lastFailureAt: nil,
            consecutiveActivationFailures: 0,
            updatedAt: now
        )
        mutate(&record)
        record.updatedAt = now
        guard Self.valid(record) else { throw PocketAppHealthError.invalid }
        try write(record)
    }

    private func read(packageID: String) throws -> Record? {
        guard Self.validPackageID(packageID) else { throw PocketAppHealthError.invalid }
        try requireSafeRoot()
        let url = recordURL(packageID: packageID)
        guard FileManager.default.fileExists(atPath: url.path) else { return nil }
        do {
            let data = try PocketAppFileSnapshot.readFileNoFollow(
                rootDirectory: rootDirectory,
                relativePath: "\(packageID).json",
                maximumBytes: Self.maximumRecordBytes
            )
            let decoder = JSONDecoder()
            decoder.dateDecodingStrategy = .iso8601
            let record = try decoder.decode(Record.self, from: data)
            guard Self.valid(record), record.packageID == packageID else {
                throw PocketAppHealthError.invalid
            }
            return record
        } catch let error as PocketAppHealthError {
            throw error
        } catch {
            throw PocketAppHealthError.invalid
        }
    }

    private func write(_ record: Record) throws {
        try requireSafeRoot()
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.sortedKeys]
        let data = try encoder.encode(record)
        guard data.count <= Self.maximumRecordBytes else { throw PocketAppHealthError.invalid }

        let temporary = rootDirectory.appendingPathComponent(".health-\(UUID().uuidString.lowercased()).tmp")
        let destination = recordURL(packageID: record.packageID)
        do {
            try data.write(to: temporary, options: [.withoutOverwriting])
            try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: temporary.path)
            guard Darwin.rename(temporary.path, destination.path) == 0 else {
                throw PocketAppHealthError.storage
            }
            let observed = try PocketAppFileSnapshot.readFileNoFollow(
                rootDirectory: rootDirectory,
                relativePath: "\(record.packageID).json",
                maximumBytes: Self.maximumRecordBytes
            )
            guard observed == data else { throw PocketAppHealthError.readback }
        } catch let error as PocketAppHealthError {
            try? FileManager.default.removeItem(at: temporary)
            throw error
        } catch {
            try? FileManager.default.removeItem(at: temporary)
            throw PocketAppHealthError.storage
        }
    }

    private func requireSafeRoot() throws {
        let values = try rootDirectory.resourceValues(forKeys: [.isDirectoryKey, .isSymbolicLinkKey])
        guard values.isDirectory == true, values.isSymbolicLink != true else {
            throw PocketAppHealthError.storage
        }
    }

    private func recordURL(packageID: String) -> URL {
        rootDirectory.appendingPathComponent("\(packageID).json", isDirectory: false)
    }

    private static func snapshot(
        package: PocketAppManagedPackage,
        record: Record?,
        now: Date
    ) -> PocketAppHealthSnapshot {
        if package.state == .disabled {
            return PocketAppHealthSnapshot(
                packageID: package.packageID,
                status: .disabled,
                reasonCode: "APP_DISABLED",
                lastUsedAt: record?.lastUsedAt,
                lastSuccessfulActivationAt: record?.lastSuccessfulActivationAt,
                consecutiveActivationFailures: record?.consecutiveActivationFailures ?? 0,
                disableSuggested: false
            )
        }
        if let record, record.consecutiveActivationFailures >= 3 {
            return PocketAppHealthSnapshot(
                packageID: package.packageID,
                status: .attention,
                reasonCode: "ACTIVATION_FAILURES",
                lastUsedAt: record.lastUsedAt,
                lastSuccessfulActivationAt: record.lastSuccessfulActivationAt,
                consecutiveActivationFailures: record.consecutiveActivationFailures,
                disableSuggested: false
            )
        }
        let inactivityReference = record?.lastUsedAt ?? record?.firstActivatedAt
        let unused = inactivityReference.map {
            now.timeIntervalSince($0) >= Self.unusedInterval
        } ?? false
        return PocketAppHealthSnapshot(
            packageID: package.packageID,
            status: unused ? .unused : .healthy,
            reasonCode: unused ? "UNUSED_30_DAYS" : "HEALTHY",
            lastUsedAt: record?.lastUsedAt,
            lastSuccessfulActivationAt: record?.lastSuccessfulActivationAt,
            consecutiveActivationFailures: record?.consecutiveActivationFailures ?? 0,
            disableSuggested: unused
        )
    }

    private static func valid(_ record: Record) -> Bool {
        let dates = [
            record.firstActivatedAt,
            record.lastSuccessfulActivationAt,
            record.lastUsedAt,
            record.lastFailureAt
        ].compactMap { $0 }
        guard record.recordVersion == 1,
              validPackageID(record.packageID),
              (0...1_000).contains(record.consecutiveActivationFailures),
              dates.allSatisfy({ $0.timeIntervalSince1970.isFinite && $0 <= record.updatedAt }) else {
            return false
        }
        if let first = record.firstActivatedAt,
           let success = record.lastSuccessfulActivationAt,
           success < first {
            return false
        }
        return true
    }

    private static func validPackageID(_ value: String) -> Bool {
        value.unicodeScalars.count <= 160
            && value.range(
                of: "^[a-z][a-z0-9]*(?:\\.[a-z0-9][a-z0-9-]*){2,}$",
                options: .regularExpression
            ) != nil
    }
}

File: windows/src/HoverPocket.Shell/PocketApps/PocketAppHealthStore.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace HoverPocket.Shell.PocketApps;

internal enum PocketAppHealthStatus
{
    Healthy,
    Attention,
    Unused,
    Disabled
}

internal sealed record PocketAppHealthSnapshot(
    string PackageId,
    PocketAppHealthStatus Status,
    string ReasonCode,
    DateTimeOffset? LastUsedAt,
    DateTimeOffset? LastSuccessfulActivationAt,
    int ConsecutiveActivationFailures,
    bool DisableSuggested);

internal sealed class PocketAppHealthException(string code) : Exception(code)
{
    internal string Code { get; } = code;
}

internal sealed class PocketAppHealthStore
{
    private sealed record HealthRecord(
        int RecordVersion,
        string PackageId,
        DateTimeOffset? FirstActivatedAt,
        DateTimeOffset? LastSuccessfulActivationAt,
        DateTimeOffset? LastUsedAt,
        DateTimeOffset? LastFailureAt,
        int ConsecutiveActivationFailures,
        DateTimeOffset UpdatedAt);

    internal static readonly TimeSpan UnusedInterval = TimeSpan.FromDays(30);
    private static readonly TimeSpan UsageWriteInterval = TimeSpan.FromMinutes(5);
    private const int MaximumRecordBytes = 16 * 1024;
    private readonly string _rootDirectory;
    private readonly JsonSerializerOptions _jsonOptions = new()
    {
        PropertyNameCaseInsensitive = false,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        UnmappedMemberHandling = JsonUnmappedMemberHandling.Disallow,
        MaxDepth = 8,
        WriteIndented = false
    };

    internal PocketAppHealthStore(string rootDirectory)
    {
        _rootDirectory = Path.GetFullPath(rootDirectory);
        try
        {
            Directory.CreateDirectory(_rootDirectory);
            RequireSafeRoot();
        }
        catch (PocketAppHealthException)
        {
            throw;
        }
        catch
        {
            throw Failure("HEALTH_STORAGE_FAILED");
        }
    }

    internal void RecordActivationSuccess(string packageId, DateTimeOffset? now = null)
    {
        var timestamp = now ?? DateTimeOffset.UtcNow;
        Update(packageId, timestamp, record => record with
        {
            FirstActivatedAt = record.FirstActivatedAt ?? timestamp,
            LastSuccessfulActivationAt = timestamp,
            ConsecutiveActivationFailures = 0
        });
    }

    internal void RecordActivationFailure(string packageId, DateTimeOffset? now = null)
    {
        var timestamp = now ?? DateTimeOffset.UtcNow;
        Update(packageId, timestamp, record => record with
        {
            LastFailureAt = timestamp,
            ConsecutiveActivationFailures = Math.Min(record.ConsecutiveActivationFailures + 1, 1_000)
        });
    }

    internal void RecordUse(string packageId, DateTimeOffset? now = null)
    {
        var timestamp = now ?? DateTimeOffset.UtcNow;
        if (!ValidPackageId(packageId)) { throw Failure("HEALTH_INVALID"); }
        var current = Read(packageId);
        if (current?.LastUsedAt is { } lastUsed
            && timestamp >= lastUsed
            && timestamp - lastUsed < UsageWriteInterval)
        {
            return;
        }
        Update(packageId, timestamp, record => record with
        {
            FirstActivatedAt = record.FirstActivatedAt ?? timestamp,
            LastUsedAt = timestamp
        });
    }

    internal IReadOnlyList<PocketAppHealthSnapshot> Snapshots(
        IReadOnlyList<PocketAppManagedPackage> packages,
        IReadOnlyList<PocketAppManagementIssue> issues,
        DateTimeOffset? now = null)
    {
        var timestamp = now ?? DateTimeOffset.UtcNow;
        var result = new List<PocketAppHealthSnapshot>();
        foreach (var package in packages.Where(item => item.State != PocketAppLifecycleState.Removed))
        {
            try
            {
                result.Add(Snapshot(package, Read(package.PackageId), timestamp));
            }
            catch
            {
                result.Add(new PocketAppHealthSnapshot(
                    package.PackageId,
                    PocketAppHealthStatus.Attention,
                    "HEALTH_METADATA_CORRUPT",
                    null,
                    null,
                    0,
                    false));
            }
        }
        foreach (var issue in issues)
        {
            HealthRecord? record = null;
            try { record = Read(issue.PackageId); } catch { }
            result.Add(new PocketAppHealthSnapshot(
                issue.PackageId,
                PocketAppHealthStatus.Attention,
                issue.ErrorCode,
                record?.LastUsedAt,
                record?.LastSuccessfulActivationAt,
                record?.ConsecutiveActivationFailures ?? 0,
                false));
        }
        return result.OrderBy(item => item.PackageId, StringComparer.Ordinal).ToArray();
    }

    private void Update(
        string packageId,
        DateTimeOffset now,
        Func<HealthRecord, HealthRecord> mutate)
    {
        if (!ValidPackageId(packageId)) { throw Failure("HEALTH_INVALID"); }
        var record = Read(packageId) ?? new HealthRecord(
            1,
            packageId,
            null,
            null,
            null,
            null,
            0,
            now);
        record = mutate(record) with { UpdatedAt = now };
        if (!Valid(record)) { throw Failure("HEALTH_INVALID"); }
        Write(record);
    }

    private HealthRecord? Read(string packageId)
    {
        if (!ValidPackageId(packageId)) { throw Failure("HEALTH_INVALID"); }
        RequireSafeRoot();
        var path = RecordPath(packageId);
        if (!File.Exists(path)) { return null; }
        try
        {
            var data = PocketAppFileSnapshot.ReadFileNoFollow(
                _rootDirectory,
                $"{packageId}.json",
                MaximumRecordBytes);
            var record = JsonSerializer.Deserialize<HealthRecord>(data, _jsonOptions)
                ?? throw Failure("HEALTH_INVALID");
            if (!Valid(record) || record.PackageId != packageId)
            {
                throw Failure("HEALTH_INVALID");
            }
            return record;
        }
        catch (PocketAppHealthException)
        {
            throw;
        }
        catch
        {
            throw Failure("HEALTH_INVALID");
        }
    }

    private void Write(HealthRecord record)
    {
        RequireSafeRoot();
        var data = JsonSerializer.SerializeToUtf8Bytes(record, _jsonOptions);
        if (data.Length > MaximumRecordBytes) { throw Failure("HEALTH_INVALID"); }
        var temporary = Path.Combine(_rootDirectory, $".health-{Guid.NewGuid():N}.tmp");
        var destination = RecordPath(record.PackageId);
        try
        {
            using (var stream = new FileStream(
                temporary,
                FileMode.CreateNew,
                FileAccess.Write,
                FileShare.None,
                4_096,
                FileOptions.WriteThrough))
            {
                stream.Write(data);
                stream.Flush(flushToDisk: true);
            }
            if (File.Exists(destination))
            {
                var attributes = File.GetAttributes(destination);
                if (attributes.HasFlag(FileAttributes.ReparsePoint)
                    || attributes.HasFlag(FileAttributes.Directory))
                {
                    throw Failure("HEALTH_STORAGE_FAILED");
                }
            }
            File.Move(temporary, destination, overwrite: true);
            var observed = PocketAppFileSnapshot.ReadFileNoFollow(
                _rootDirectory,
                $"{record.PackageId}.json",
                MaximumRecordBytes);
            if (!observed.AsSpan().SequenceEqual(data))
            {
                throw Failure("HEALTH_READBACK_FAILED");
            }
        }
        catch (PocketAppHealthException)
        {
            try { if (File.Exists(temporary)) { File.Delete(temporary); } } catch { }
            throw;
        }
        catch
        {
            try { if (File.Exists(temporary)) { File.Delete(temporary); } } catch { }
            throw Failure("HEALTH_STORAGE_FAILED");
        }
    }

    private void RequireSafeRoot()
    {
        var attributes = File.GetAttributes(_rootDirectory);
        if (!attributes.HasFlag(FileAttributes.Directory)
            || attributes.HasFlag(FileAttributes.ReparsePoint))
        {
            throw Failure("HEALTH_STORAGE_FAILED");
        }
    }

    private string RecordPath(string packageId) => Path.Combine(_rootDirectory, $"{packageId}.json");

    private static PocketAppHealthSnapshot Snapshot(
        PocketAppManagedPackage package,
        HealthRecord? record,
        DateTimeOffset now)
    {
        if (package.State == PocketAppLifecycleState.Disabled)
        {
            return new PocketAppHealthSnapshot(
                package.PackageId,
                PocketAppHealthStatus.Disabled,
                "APP_DISABLED",
                record?.LastUsedAt,
                record?.LastSuccessfulActivationAt,
                record?.ConsecutiveActivationFailures ?? 0,
                false);
        }
        if (record is { ConsecutiveActivationFailures: >= 3 } failed)
        {
            return new PocketAppHealthSnapshot(
                package.PackageId,
                PocketAppHealthStatus.Attention,
                "ACTIVATION_FAILURES",
                failed.LastUsedAt,
                failed.LastSuccessfulActivationAt,
                failed.ConsecutiveActivationFailures,
                false);
        }
        var inactivityReference = record?.LastUsedAt ?? record?.FirstActivatedAt;
        var unused = inactivityReference is { } reference && now - reference >= UnusedInterval;
        return new PocketAppHealthSnapshot(
            package.PackageId,
            unused ? PocketAppHealthStatus.Unused : PocketAppHealthStatus.Healthy,
            unused ? "UNUSED_30_DAYS" : "HEALTHY",
            record?.LastUsedAt,
            record?.LastSuccessfulActivationAt,
            record?.ConsecutiveActivationFailures ?? 0,
            unused);
    }

    private static bool Valid(HealthRecord record)
    {
        var dates = new[]
        {
            record.FirstActivatedAt,
            record.LastSuccessfulActivationAt,
            record.LastUsedAt,
            record.LastFailureAt
        }.Where(value => value is not null).Select(value => value!.Value);
        return record.RecordVersion == 1
            && ValidPackageId(record.PackageId)
            && record.ConsecutiveActivationFailures is >= 0 and <= 1_000
            && dates.All(value => value <= record.UpdatedAt)
            && !(record.FirstActivatedAt is { } first
                && record.LastSuccessfulActivationAt is { } success
                && success < first);
    }

    private static bool ValidPackageId(string value) =>
        value.Length <= 160
        && System.Text.RegularExpressions.Regex.IsMatch(
            value,
            "^[a-z][a-z0-9]*(?:\\.[a-z0-9][a-z0-9-]*){2,}$",
            System.Text.RegularExpressions.RegexOptions.CultureInvariant);

    private static PocketAppHealthException Failure(string code) => new(code);
}
